### PR TITLE
Create 005-javascript-style.md

### PR DESCRIPTION
**Pull request by schalkneethling**
Document Javascript code style guidelines and Use ESlint and Prettier across MDN

First stab at putting together an ADR for introducing a JavaScript code style guide as well as related tools. Please let me know your thoughts @peterbe @escattone @davidflanagan

### CODE REVIEW

**Comment by schalkneethling** [comment]
Not sure if this should be added to the ADR or not so, adding it as a comment:

Prettier config is .prettierrc

```
{}
```

ESLint config is as follows:

```
module.exports = {
    env: {
        browser: true,
        es6: true
    },
    extends: 'eslint:recommended',
    parserOptions: {
        ecmaVersion: 8
    },
    rules: {
        'no-global-assign': 2,
        indent: [2, 4, { SwitchCase: 1 }],
        'linebreak-style': [2, 'unix'],
        quotes: [2, 'single'],
        semi: [2, 'always'],
        curly: [2, 'all'],
        camelcase: [
            2,
            {
                properties: 'always'
            }
        ],
        eqeqeq: [2, 'smart'],
        'no-console': [
            'error',
            {
                allow: ['error', 'info']
            }
        ],
        'one-var-declaration-per-line': [2, 'always'],
        'new-cap': 2
    },
    globals: {
        console: true,
        expect: true,
        exports: true,
        module: true,
        require: true,
        test: true
    }
};
```

For React based projects:

{
    "extends": [
      "eslint:recommended",
      "plugin:import/errors",
      "plugin:react/recommended",
      "plugin:jsx-a11y/recommended",
      "prettier",
      "prettier/react"
    ],
    "rules": {
      "react/prop-types": 0,
      "no-console": 1,
      "react-hooks/rules-of-hooks": "error",
      "react-hooks/exhaustive-deps": "warn"
    },
    "plugins": ["react", "react-hooks", "import", "jsx-a11y"],
    "parser": "babel-eslint",
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module",
      "ecmaFeatures": {
        "jsx": true
      }
    },
    "env": {
      "browser": true,
      "es6": true,
      "jest": true,
      "node": true
    }
  }

**Comment by escattone** [comment]
👍 I really like the idea of going with something like `prettier` and `eslint`. For me, the code standardization/consistency and basic checking outweigh any quibbles I may have with the choices the tools make. I'll defer to those who have coded in JavaScript more than I have for the exact configuration settings.

File: ADRs/005-javascript-style.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Document Javascript code style guidelines and Use ESlint and Prettier across MDN
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]:
```suggestion
Document JavaScript code style guidelines and Use ESlint and Prettier across MDN
```

Comment (peterbe) [reply ADRs/005-javascript-style.md]:
```suggestion
Document Javascript code style guidelines and use ESlint and Prettier across MDN
```

Comment (peterbe) [reply ADRs/005-javascript-style.md]: I don't think "Document" is the right word.

**Comment by peterbe** [comment]
I might be pedantic but I think the formatting should be separated from the code-quality. 

ESLint can do both formatting and code-quality and I'm not interested in ESLint formatting warnings/errors because they can be automated away anyway. I want me ESling to tell me about unused variables or unreachable code. 

As David pointed out, this is a great resource for setting up a config where you get eslint and prettier to work together nicely: https://prettier.io/docs/en/integrating-with-linters.html

But for now I'd like the ADR to be smaller in scope and just worry itself about formatting. 

Also, I think the ADR should include the proposed default Prettier config. I.e. the ADR should be:

* That all projects format with Prettier. 

* That the default Prettier config is: ....


...but phrased more verbosely and distinctly. 

Also, I think the ADR could include a sentence about this being the **default** and that deviation is accepted if there exists legacy code that would be too re-formatted if adopting the default standard.

**Comment by ddbeck** [comment]
Thanks for starting this PR, @schalkneethling! I have a bunch of thoughts on this, but for the most part I'm trying to wear my BCD maintainer hat for this (i.e., asking myself, what would I need to know to follow this recommendation or to accept a PR that purports to follow this recommendation?).

Here are my first impressions of the PR as it stands:

* The ADR should contain the recommended config. There should be one place for someone starting a new MDN project to find the recommended style and it may as well be the ADR itself. (From here on, I'm going to assume that the proposed config(s) get added to the ADR.)
* The ADR ought to drop a linting recommendation with the exception of how to prevent ESLint and prettier from clashing (more on that in a moment). I happen to like ESLint (I've only ever used it for style and linting; I'm new to prettier), but I think it's preferable to [publish an MDN preset package](https://eslint.org/docs/developer-guide/shareable-configs) rather than provide a configuration through the ADR.
* Why does the non-React config not extend prettier's ESLint config ([as is apparently recommended](https://prettier.io/docs/en/integrating-with-linters.html#eslint))? The React config does extend it and I'm not sure I understand why the two are different in this respect. 
* The React config implies that the ADR proposes adopting a style guide for JSX; the title and introduction of the ADR should reflect that, if that's your intent.
* The "Alternatives" section is a bit underdeveloped. One big alternative is to use ESLint style rules. Some questions I'd expect an alternatives section to answer: Why not adopt some existing ESLint style rules or a config wholesale (e.g., Airbnb, Standard, Google)? What makes using two tools preferable to one?

There are a few more things that I would hope to see in a complete ADR:

* I'd like to see a version for prettier specified. [According to the prettier docs](https://prettier.io/docs/en/install.html), "We recommend pinning an exact version of prettier in your `package.json` as we introduce stylistic changes in patch releases." Given prettier's versioning policy, I don't trust that the ADR actually proposes a shared style, absent a version number.
* As I've mentioned elsewhere, I'm not enamored with the fact that prettier doesn't specify its own behavior, but if we're going to offload style decisions to prettier, then the ADR should link to [the page that partially describes prettier's style](https://prettier.io/docs/en/rationale.html).
* The ADR advises "adopting" and "using" prettier; I'd like to some direction about what's meant by that in practice (e.g., a canonical name (and invocation) for a node package script). If we're offloading style to a tool, then invocation of the tool _is_ our style.

**Comment by schalkneethling** [comment]
> Why does the non-React config not extend prettier's ESLint config (as is apparently recommended)? The React config does extend it and I'm not sure I understand why the two are different in this respect.

That was merely an oversight on my part, thanks for pointing that out @ddbeck

**Comment by schalkneethling** [comment]
> The "Alternatives" section is a bit underdeveloped. One big alternative is to use ESLint style rules. Some questions I'd expect an alternatives section to answer: Why not adopt some existing ESLint style rules or a config wholesale (e.g., Airbnb, Standard, Google)? What makes using two tools preferable to one?

I know this peaks to JavaScript specifically but, thinking ahead, Prettier supports many more languages other than JS and would be used for those as well. It, therefore, makes sense to use one tool for all formatting.

I do need to add this to the "Alternatives" section though.

**Comment by schalkneethling** [comment]
> I'd like to see a version for prettier specified. According to the prettier docs, "We recommend pinning an exact version of prettier in your package.json as we introduce stylistic changes in patch releases." Given prettier's versioning policy, I don't trust that the ADR actually proposes a shared style, absent a version number.

I agree and would like it if we generally always do `npm i [-D] --save-exact ...` for most(all) things.

**Comment by schalkneethling** [comment]
I have updated the PR/ADR to focus on JavaScript formatting using Prettier.

**Comment by peterbe** [comment]
@ddbeck ...

> If we're offloading style to a tool, then invocation of the tool is our style.

If I parse that sentence correctly, I wholeheartedly agree. I think what this ADR is trying to say is "Instead of a specified formatting specification, the formatting is whatever Prettier provides (with our (if any) custom overrides)"

Let's supersede any formatting of our own. If it passes Prettier, it's formatted correctly.

**Comment by peterbe** [comment]
On the topic of exact versions, first things first; Our `package.json` should not have ANY ranges or regexes. Only exact versions. Both for `dependencies` and for `devDependencies`. Generally a lock file solves this but I'm hazy on how npm and `package-lock.json` works. Ideally, the package-lock.json should dictate which exact packages get installed and essentially the `package.json` should not matter at all. Not in local dev or in CI. ...in terms of dependencies. 

I think it's unrelated to this ADR but if time allows I'd like to remove all the little `^` prefixes on the version numbers in kuma's `package.json`.

**Comment by schalkneethling** [comment]
> Let's supersede any formatting of our own. If it passes Prettier, it's formatted correctly.

+1

--

I also agree that a discussion on module versions and using `--save-exact` is not in scope for this discussion. While it is important and needs doing, we will have to have the discussion separately.

**Comment by ddbeck** [comment]
@peterbe Yeah, that's pretty much it. I was trying to make explicit that our style is in the invocation of prettier, where "invocation" is meant to include all of these things: the version of prettier (because prettier can change its style at any version—so consistency across projects means that you have the same prettier version across projects), the config for prettier, and the specific command-line options

File: ADRs/005-javascript-style.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+Use Prettier for JavaScript code formatting across all MDN codebases.
+
+### Context
+
+We have a number of repositories across the MDN organization that uses JavaScript. To ease 
+contributions, shorten pull requests review time, and ease transition between projects,
+it is suggested that we adopt Prettier across all projects inside the MDN org for JavaScript code formatting.
+
+To ensure consistency, it is also suggested that we utilise Husky[] and enable a pre-commit hook to auto format all JavaScript.
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]: I think we should remove this. It's an implementation detail. It's not unimportant but it's too early to bring up that kind of detail.

File: ADRs/005-javascript-style.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  "format": "prettier --write \"src/**/*.{js,jsx}\"",
+  "format:check": "prettier --list-different \"src/**/*.{js,jsx}\"",
+},
+"husky": {
+  "hooks": {
+    "pre-commit": "npm format"
+  }
+}
+```
+
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]: I would remove all of this and replace it with a simple and bold statement that says that our preferred configuration is that which is the current default in Prettier as per whatever is the latest version. 

A fair concern that @ddbeck brought up is that we can "outsource" this opinion to Prettier but Prettier is potentially going to change its implementation and if we *don't* say "Preferred configuration is that of Prettier version 1.18.2" then, code bases will "break" when new versions of Prettier comes out. 

Granted, we *could* say a specific version of Prettier, and imply its default configuration, and then update this ADR when a new version comes out.

File: ADRs/005-javascript-style.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+### Consequences
+
+We have standard JavaScript code formatting across projects, and tools that enforce these standards. 
+
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]: Please add something like:

Code formatting is no longer manually maintained or reviewed. CI is used to verify strict adherence to this rather than human intervention. Failure to format code correctly, with Prettier, in pull requests should break and put the onus on the contributor to fix linting before commencing human code review.

File: ADRs/005-javascript-style.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+### Alternatives Considered
+
+This is very norrowly scoped and suggests a tool that is standard across the industry, and already used on projects such as Kuma, so other options have not been explored.
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]: I don't know where else to mention it but the ADR should mention that individual projects are allowed to deviate if there are solid reasons to do so. For example `kuma` already uses 4-spaces indentation and I think it should allow continuing to do so (to avoid crazy-big changes that ruin too much of the git history).
It should also mention this rule is only allowed for existing projects with a significant amount of JavaScript code. 

Mentioning this means an "alternative considered" is 4-spaces indentation but still with Prettier.

File: ADRs/005-javascript-style.md (hunk 3 above)
Comment (peterbe) [reply ADRs/005-javascript-style.md]: Actually, I think we should be explicit and say, in this ADR, that the use of Prettier does *not* dictate linting or tooling. I.e. that nowhere here does it say that you should use `eslint-prettier-react` or something like that.

File: ADRs/005-javascript-style.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+Javascript code formatting with Prettier across MDN
```
Comment (peterbe) [inline ADRs/005-javascript-style.md]:
```suggestion
JavaScript code formatting with Prettier across MDN repositories
```

**Comment by schalkneethling** [comment]
@peterbe Updates made to the copy. Thank you for the feedback.

File: ADRs/005-javascript-style.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+### Context
+
+We have a number of repositories across the MDN organization that uses JavaScript. To ease 
+contributions, shorten pull requests review time, and ease transition between projects,
+it is suggested that we adopt Prettier across all projects inside the MDN org for JavaScript code formatting.
+
+### Configuration
+
+Configuration will be based on Prettier's default configuration at the time this ADR is excepted. Currently the latest version of Prettier is 1.18.2
```
Comment (ddbeck) [inline ADRs/005-javascript-style.md]: My only note on the current draft, there's a little typo to fix:

```suggestion
Configuration will be based on Prettier's default configuration at the time this ADR is accepted. Currently the latest version of Prettier is 1.18.2
```

Comment (peterbe) [inline ADRs/005-javascript-style.md]: I'm still not sure about this. In particular, to mention a specific version. I think that's up to the individual MDN projects to worry about. For this ADR, the **most important** point of this whole ADR is...

* That we *use* Prettier
* That *our* configuration override(s) is exactly *nothing* (aka. `{}`)

Suppose that we deflect, from all specific opinions about JavaScript styles, to Prettier; our preferences isn't what Prettier *says* but that we **use** Prettier. The only advantage of sticking to a specific version is that we can say, for example, "After foo operator is 1 empty space followed by a `{` yada yada". But who cares?! Every single JavaScript project we have will have a *version* of `prettier` in its `package.json`. That's what matters. If someone makes a pull request, using the wrong/different version of Prettier on their laptop, it should break in CI. The right version is whatever is in the `package.json` and projects should be allowed (and highly recommended) to keep updating this. 

In conclusion, I suggest that we change this sentence to...

```suggestion
Our configuration is the default configuration in Prettier. I.e. `{}`. 
```

Comment (peterbe) [reply ADRs/005-javascript-style.md]: My point is that you don't need to look up specifics or even need to know them because it's entirely work taken over my machines. 

Personal preferences and manual labor is a thing of the past.

Comment (schalkneethling) [reply ADRs/005-javascript-style.md]: Gotcha, I misunderstood your previous comment.

**Comment by schalkneethling** [comment]
@ddbeck @escattone @Elchi3 @chrisdavidmills Your thoughts? 

@peterbe I also reckon we need to define next steps. What does merged/accepting the ADR mean for projects on /mdn/?

**Comment by chrisdavidmills** [comment]
I am also happy with doing this; Prettier seems popular and well-supported, and it works well.

**Comment by peterbe** [comment]
> What does merged/accepting the ADR mean for projects on /mdn/?

For one/two things at least these can get finished:

* https://github.com/mdn/stumptown-renderer/pull/31
* https://github.com/mdn/stumptown-content/pull/28

All the new JS code in Kuma is already pretty. I doubt we care about the old "wiki JS code" in Kuma. I.e. care enough to mess up its git blame history too much.

**Comment by ddbeck** [comment]
I'm a bit disappointed that details have fallen out of the ADR, particularly the version number for prettier. The ADR, as written, doesn't do much to solve the problems I have on BCD. I know now that I have to choose prettier and that's it. I'll still have to make judgement calls about what version, when (if ever) to upgrade, and I won't necessarily get the same style as other MDN projects unless I coordinate with them independently. And because there's no narrative about why prettier was chosen—no alternatives were considered and that decision itself is unexplained—I don't have any clues as to what sort of motivations and judgements went into the selection of prettier specifically (y'all seem to like prettier because *\<waves hands furiously>*). I'll have to make those judgement calls on my own whenever they come up.

That said it's clear that I've been running against the grain on this stuff. If I'm genuinely the only person who cares, then don't me be a blocker. Let this be my last word on the matter and move on.

**Comment by schalkneethling** [comment]
> And because there's no narrative about why prettier was chosen—no alternatives were considered and that decision itself is unexplained—I don't have any clues as to what sort of motivations and judgements went into the selection of prettier

@ddbeck Are there alternatives you would suggest we consider?

**Comment by ddbeck** [comment]
I'm not really advocating for any other tools, but I know there are a bunch of ESLint presets for various style guides (e.g., Standard, Google, Airbnb — Will and I used airbnb-base on mdn/short-descriptions). I went on a bit about this alternatives question previously: https://github.com/mdn/mdn/pull/73#issuecomment-498216474. But to be clear, what I'm aiming for, in the absence of explicit directives about what to use and how, is information that helps me make decisions that are in line with the motivations and intentions of the ADR.

**Comment by peterbe** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1561435

mozilla-central is switching to Prettier. On my phone right now so I can't look up what overrides they chose but sure is more reason to switch to Prettier.

**Comment by schalkneethling** [comment]
> [bugzilla.mozilla.org/show_bug.cgi?id=1561435](https://bugzilla.mozilla.org/show_bug.cgi?id=1561435)
> 
> mozilla-central is switching to Prettier. On my phone right now so I can't look up what overrides they chose but sure is more reason to switch to Prettier.

From looking at the tracker bug it seems like they are going to format everything using Prettier. If that is the case, I am thinking we should do the same even if it might dirty the history a bit for a while.

**Comment by Elchi3** [comment]
More info on the m-c prettierify: https://docs.google.com/document/d/1UDERMflocqdszMGhhtxiVhaCTVOHo6jxLsGbt4BR9uw/edit

This seems to be the config: https://dxr.mozilla.org/mozilla-central/source/.prettierrc

**Comment by karlhorky** [comment]
@peterbe mentioned this comment re: single quotes (`"singleQuote": true,` in `.prettierrc`) from mdn/stumptown-content#28 would be relevant for this thread:

---

> I'm very happy for us to use prettier, but is there a reason we using double quotes for strings? This is contrary to our practice on any other MDN project that I know of.

This is a common change to the default config.

There is a big discussion around changing the default to single quotes (albeit with no resolution yet).

https://github.com/prettier/prettier/issues/4102

But with some positive signals from the maintainers and creator ("we always meant to do that"):

https://github.com/prettier/prettier/issues/3503#issuecomment-352802184

**Comment by karlhorky** [comment]
Maybe also [this one](https://github.com/mdn/mdn/issues/71#issuecomment-508459197), from #71.

---

I for one would suggest the approach (inspired by that of `create-react-app`):

- [ESLint with only error-avoiding rules enabled](https://github.com/facebook/create-react-app/blob/24489ac0a667af416f1d59dd806dfc2623aabe36/packages/eslint-config-react-app/index.js) (no stylistic rules)
- Prettier for any code style rules

**Comment by peterbe** [comment]
I think @karlhorky you're comment kinda sums up what's going on here with this PR and why Schalk didn't attempt to re-work https://github.com/mdn/mdn/issues/71
I.e. *this* PR attempts to keep things smaller and simpler by exclusively focussing on formatting.

**Comment by vinyldarkscratch** [comment]
A little update on this: I found out that `mozilla-central` [has a Prettier config already](https://dxr.mozilla.org/mozilla-central/source/.prettierrc).  (It also has an ESLint plugin, apparently, which integrates with Prettier.)  Perhaps we should use that for MDN as well?
